src/sentry/eventstream/errors/backend.py

```diff
@@ -0,0 +1,323 @@
+
+    def end_delete_tag(self, state):
+        state = state.copy()
+        state["datetime"] = datetime.now(tz=pytz.utc)
+        self._backend.send(
+            state["project_id"], "end_delete_tag", extra_data=(state,), asynchronous=False
+        )
+
+    def tombstone_events_unsafe(
+        self, project_id, event_ids, old_primary_hash=False, from_timestamp=None, to_timestamp=None
+    ):
+        """
+        Tell Snuba to eventually delete these events.
+
+        This marks events as deleted but does not immediately exclude those
+        events from all queries. Because of that limitation this is not proper,
+        because not immediate, event deletion.
+
+        "Proper" group deletion is essentially running this function for every
+        event in the group, plus `exclude_groups` to make sure the changes are
+        immediately user-visible.
+
+        Reprocessing (v2) splits a group into events-to-be-reprocessed
+        (re-insert with new group_id) and events-to-be-deleted
+        (`tombstone_events`), then excludes the group from all queries
+        (`exclude_groups`).
+
+        :param old_primary_hash: If present, the event is only tombstoned
+            to be reinserted over with a guaranteed-different primary hash.
+            This is necessary with Snuba's errors table as the primary_hash is
+            part of the PK/sortkey.
+        """
+
+        state = {
+            "project_id": project_id,
+            "event_ids": event_ids,
+            "old_primary_hash": old_primary_hash,
+            "from_timestamp": from_timestamp,
+            "to_timestamp": to_timestamp,
+        }
+        self._backend.send(project_id, "tombstone_events", extra_data=(state,), asynchronous=False)
+
+    def replace_group_unsafe(
+        self, project_id, event_ids, new_group_id, from_timestamp=None, to_timestamp=None
+    ):
+        """
+        Tell Snuba to move events into a new group ID
+
+        Same caveats as tombstone_events
+        """
+
+        state = {
+            "project_id": project_id,
+            "event_ids": event_ids,
+            "new_group_id": new_group_id,
+            "from_timestamp": from_timestamp,
+            "to_timestamp": to_timestamp,
+        }
+        self._backend.send(project_id, "replace_group", extra_data=(state,), asynchronous=False)
+
+    def exclude_groups(self, project_id, group_ids):
+        """
+        Exclude a group from queries for a while until event tombstoning takes
+        effect. See docstring of `tombstone_events`.
+
+        `exclude_groups` basically makes Snuba add `where group_id not in (1,
+        2, ...)` to every query.
+        """
+        state = {"project_id": project_id, "group_ids": group_ids}
+        self._backend.send(project_id, "exclude_groups", extra_data=(state,), asynchronous=False)
+
+
+class SnubaErrorsEventStreamAPI(ErrorsEventStreamAPI):
+    def __init__(self):
+        self._backend = SnubaEventStreamBackend("events")
```

Comment:
You should not need this level of hierarchy. 
You already solved the problem of having multiple backends by composition by moving all the implementation specific logic in the `EventStreamBackend`.

If you add a subclass here you are stuck again with multi dimensional hierarchies. 
Your problem here is likely the instantiation of the right backend. In general you have two possible solutions:
- make the backend know what to instantiate (with some sort of Factory pattern)
- make it a problem of whoever creates the `ErrorsEventStreamAPI` object to also instantiate the backend and pass the right one to the init.

There are often cases for either of the solutions. In our code base we have a specific pattern and abstraction for the first: `ServiceDelegator` You can create a service and the settings decide what you will instantiate. This should work well here as it is through settings that you decide whether to use Snuba or Kafka.